### ISSUE TITLE
Add `CreateReply::set_embed` API

### ISSUE THREAD

**Issue by s0lst1ce** [kangalioo/poise#27]
Serenity's `CreateMessage` struct has the `set_embed(embed: CreateEmbed)` API when poise only has `embed(embed: &mut CreateEmbed)`. While taking a mutable reference can be useful it also causes some lifetime issues. Consider the following code:
```rs
pub type Context<'a> = poise::Context<'a, GlobalData, BotError>;
fn some_func(ctx: Context<'_>) {
   //embed is bound to the scope of the function
    let mut embed = CreateEmbed::default();
    // but we make a ref out of it
    ctx.send(|r| r.embed(|e| &mut embed)).await;
    //then we want to send it again
    ctx.send(|r| r.embed(|e| &mut embed)).await;
```
As you can see this doesn't work because `embed` doesn't live long enough since it's bound to the scope's lifetime (not sure about the wording, I'm still pretty new to lifetimes ^^). Of course I could just move the entire embed creation process within the closure but that can be hard to follow and more importantly it means I can't just `clone()` the `CreateEmbed` if I want to send it again!
If I'm not just missing something I'd like there to be a `set_embed` API that takes an owned `CreateEmbed`.

**Comment by kangalioo** [comment]
Does this work for you?

```rust
let mut embed = CreateEmbed::default();
ctx.send(|r| r.embed(|e| {*e = embed.clone(); e}).await;
ctx.send(|r| r.embed(|e| {*e = embed.clone(); e}).await;
```

**Comment by s0lst1ce** [comment]
It does. I don't understand why though, would you mind explaining it to me? I don't understand why doing this like this doesn't cause a "embed.clone() doesn't live long enough" issue?

Also would you rather have me ask on the discord if there's a way to do X before opening an issue about it here. I feel like a few of the issues I opened where due to my lack of understanding of rust.

**Comment by kangalioo** [comment]
I don't mind the github issues, though Discord may be better just because it allows faster conversation.

> I don't understand why though, would you mind explaining it to me?

I don't mind at all :)

The signature of `CreateReply::embed` is `impl FnOnce(&mut CreateEmbed) -> &mut CreateEmbed`. Lifetimes are left off, so they're implied to be the same. Which means that the closure must return a CreateEmbed of the same lifetime (or larger) as it was given.

With `r.embed(|e| &mut embed)`, that is not the case. `&mut embed`'s lifetime is always bound to the local scope and doesn't necessarily match the input lifetime.

`r.embed(|e| {*e = embed.clone(); e}` avoids the issue, because instead of returning a reference to a local variable, it copies the contents of the local variable into the given reference.

<details>
<summary>
There would be a way to change `CreateReply::embed`'s signature to make your code work, though it has its own problems
</summary>

```diff
-pub fn embed(
+pub fn embed<'b>(
     &mut self,
-    f: impl FnOnce(&mut serenity::CreateEmbed) -> &mut serenity::CreateEmbed,
+    f: impl FnOnce(&mut serenity::CreateEmbed) -> &'b mut serenity::CreateEmbed,
 ) -> &mut Self {
```

This reads as "the closure may be supplied a CreateEmbed of arbitrary lifetime, but it can return a lifetime of its choosing". However, this becomes problematic when using the builder normally, by calling methods on it

```rust
ctx.send(|r| r.embed(|e| e.description("lorem ipsum")));
```

```rust
error: lifetime may not live long enough
  --> examples/framework_usage/main.rs:24:30
   |
24 |     ctx.send(|r| r.embed(|e| e.description("lorem ipsum")));
   |                           -- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
   |                           ||
   |                           |return type of closure is &'2 mut CreateEmbed
   |                           has type `&'1 mut CreateEmbed`
```

Recall "the closure may be supplied a CreateEmbed of arbitrary lifetime, but it can return a lifetime of its choosing": the compiler complains because what if `e`'s lifetime is smaller than the lifetime we (implicitly) chose? Then the returned value has a lifetime smaller than what it needs to have.

I hope I haven't lost you somewhere along the explanation. Anyways, this is why the API is currently the way it is.

</details>

**Comment by s0lst1ce** [comment]
This was very clear and perfectly answered my question. Thanks.
